lighthouse-core/report/v2/renderer/report-features.js

```diff
@@ -0,0 +1,276 @@
+      case 'copy':
+        this.onCopyButtonClick();
+        break;
+      case 'open-viewer':
+        this.sendJSONReport();
+        break;
+      case 'print':
+        this.expandDetailsWhenPrinting();
+        window.print();
+        break;
+      case 'save-json': {
+        const jsonStr = JSON.stringify(this.json, null, 2);
+        this._saveFile(new Blob([jsonStr], {type: 'application/json'}));
+        break;
+      }
+      case 'save-html': {
+        let htmlStr = '';
+
+        // Since Viewer generates its page HTML dynamically from report JSON,
+        // run the ReportGenerator. For everything else, the page's HTML is
+        // already the final product.
+        // if (e.target.dataset.context !== 'viewer') {
+        //   htmlStr = this._document.documentElement.outerHTML;
+        // } else {
+        //   const reportGenerator = new ReportGeneratorV2();
+        //   htmlStr = reportGenerator.generateReportHTML(this.json);
+        // }
+        // TODO: fix viewer.
+        htmlStr = this._document.documentElement.outerHTML;
+
+        try {
+          this._saveFile(new Blob([htmlStr], {type: 'text/html'}));
+        } catch (err) {
+          this.logger.error('Could not export as HTML. ' + err.message);
+        }
+        break;
+      }
+    }
+
+    this.closeExportDropdown();
+    this._document.removeEventListener('keydown', this.onKeyDown);
+  }
+
+  /**
+   * Keydown handler for the document.
+   */
+  onKeyDown(e) {
+    if (e.keyCode === 27) { // ESC
+      this.closeExportDropdown();
+    }
+  }
+
+  /**
+   * Opens a new tab to the online viewer and sends the local page's JSON results
+   * to the online viewer using postMessage.
+   */
+  sendJSONReport() {
+    const VIEWER_ORIGIN = 'https://googlechrome.github.io';
+    const VIEWER_URL = `${VIEWER_ORIGIN}/lighthouse/viewer/`;
+
+    // Chrome doesn't allow us to immediately postMessage to a popup right
+    // after it's created. Normally, we could also listen for the popup window's
+    // load event, however it is cross-domain and won't fire. Instead, listen
+    // for a message from the target app saying "I'm open".
+    window.addEventListener('message', function msgHandler(e) {
+      if (e.origin !== VIEWER_ORIGIN) {
+        return;
+      }
+
+      if (e.data.opened) {
+        popup.postMessage({lhresults: this.json}, VIEWER_ORIGIN);
+        window.removeEventListener('message', msgHandler);
+      }
+    }.bind(this));
+
+    const popup = window.open(VIEWER_URL, '_blank');
+  }
+
+  /**
+   * Expands details while user using short cut to print report
+   */
+  printShortCutDetect(e) {
+    if ((e.ctrlKey || e.metaKey) && e.keyCode === 80) { // Ctrl+P
+      this.expandDetailsWhenPrinting();
+    }
+  }
+
+  /**
+   * Expands audit `<details>` when the user prints the page.
+   * Ideally, a print stylesheet could take care of this, but CSS has no way to
+   * open a `<details>` element.
+   */
+  expandDetailsWhenPrinting() {
```

Comment:
Renamed and updated the selectors. 

The listeners need to be setup before printing happens. To further complicate things, there are two ways users can print a report: using "ctrl+p", and when click Export -> Print. The printing stuff is weird due to the browser compat. `window.addEventListener('beforeprint', this.expandAllDetails);` doesnt even work correctly in FF :(